pydoctor/astbuilder.py

```diff
@@ -153,65 +153,105 @@ def visit_ImportFrom(self, node):
+                modname = f'{parent.fullName()}.{modname}'
 
-        mod = self.system.getProcessedModule(modname)
-        if mod is not None:
-            assert mod.state in [model.ProcessingState.PROCESSING,
-                                 model.ProcessingState.PROCESSED]
-            expandName = mod.expandName
+        if node.names[0].name == '*':
+            self._importAll(modname)
         else:
-            expandName = lambda name: modname + '.' + name
+            self._importNames(modname, node.names)
+
+    def _importAll(self, modname: str) -> None:
+        """Handle a C{from <modname> import *} statement."""
+
+        mod = self.system.getProcessedModule(modname)
+        if mod is None:
+            # We don't have any information about the module, so we don't know
+            # what names to import.
+            self.builder.warning("import * from unknown", modname)
+            return
+
+        self.builder.warning("import *", modname)
+
+        # Get names to import: use __all__ if available, otherwise take all
+        # names that are not private.
+        names = mod.all
+        if names is None:
+            names = [
+                name
+                for name in chain(mod.contents.keys(),
+                                  mod._localNameToFullName_map.keys())
+                if not name.startswith('_')
+                ]
+
+        # Add imported names to our module namespace.
         _localNameToFullName = self.builder.current._localNameToFullName_map
-        for al in node.names:
-            fromname, asname = al.name, al.asname
-            if fromname == '*':
-                if mod is None:
-                    self.builder.warning("import * from unknown", modname)
-                    return
-                self.builder.warning("import *", modname)
-                if mod.all is not None:
-                    names = mod.all
-                else:
-                    names = (
-                        k
-                        for k in chain(mod.contents.keys(),
-                                       mod._localNameToFullName_map.keys())
-                        if not k.startswith('_')
-                        )
-                for n in names:
-                    _localNameToFullName[n] = expandName(n)
-                return
+        expandName = mod.expandName
+        for name in names:
+            _localNameToFullName[name] = expandName(name)
+
+    def _importNames(self, modname: str, names: Iterable[ast.alias]) -> None:
+        """Handle a C{from <modname> import <names>} statement."""
+
+        # Process the module we're importing from.
+        # If we're importing from a package, 'mod' will be the __init__ module.
+        mod = self.system.getProcessedModule(modname)
+        obj = self.system.objForFullName(modname)
+
+        # Are we importing from a package?
+        is_package = isinstance(obj, model.Package)
+        assert is_package or obj is mod or mod is None
```

Comment:
A minor question of style... do we need the assertions.
Also, for me `is_package` is a great variable name and the comment is not needed.

Reply:
I think this assertion is useful, as it makes a non-trivial statement about the code. This is illustrated by the fact that originally I forgot about the `or mod is None` part, which can happen if there is a name clash between a module and a variable in `__init__.py` within the same package.

One reason to use asserts in situations like this is that it acts as a formal and verified statement about the code. If the same statement were included as a comment, there would be more room for misinterpretation and if the statement is not or no longer true, we wouldn't know.

Another reason to use asserts is that they're active across all unit and system tests, so they help cover a much larger state space. Despite having around 90% branch coverage, there are still quite a few bugs in the `astbuilder` and `model` modules that we're not catching, because either the test cases don't do enough validations or a bug only occurs in particular combinations of states that are tested separately.

Reply:
Thanks for the info. We can keep it.

Another matter of style: Move `is_package = isinstance(obj, model.Package)` on line 250 so that it is closer to the place where it is used.

Will make the code easier to read...
I was searching for placer between line 211 and 250 to see in which other place it is used.

Reply:
Line 250 is inside the loop, while `is_package` does not depend on the thing we're iterating through. So I don't think it should be inside the loop, both for clarity and for performance.

I thought about moving the re-export code to a separate method, since that takes up the bulk of the loop's body, but doing that would create a method with 3 arguments (`mod`, `orgname` and `asname`), so it I'm not sure it would improve readability.